src/expression_builder.rs

```diff
@@ -0,0 +1,159 @@
+    instruction::Instruction,
+    operators::Operator,
+    TypeSystem,
+};
+
+#[derive(Clone)]
+pub enum Operand<TS: TypeSystem> {
+    Function {
+        addr: usize,
+        args: Vec<Operand<TS>>,
+        stack_size: usize,
+    },
+    ValueRef(usize),
+    ValueRaw(TS::Value),
+    Expression(Box<ExpressionBuilder<TS>>),
+}
+
+#[derive(Default, Clone)]
+pub struct ExpressionBuilder<TS: TypeSystem> {
+    operator: Option<Operator<TS>>,
+    operands: (Option<Operand<TS>>, Option<Operand<TS>>),
+}
+
+impl<TS: TypeSystem> ExpressionBuilder<TS> {
+    pub fn new() -> ExpressionBuilder<TS> {
+        ExpressionBuilder {
+            operator: None,
+            operands: (None, None),
+        }
+    }
+
+    pub fn set_value(&mut self, value: Operand<TS>) -> &mut ExpressionBuilder<TS> {
+        match &self.operands {
+            (None, _) => self.operands.0 = Some(value),
+            (Some(_), None) => self.operands.1 = Some(value),
+            _ => (),
+        }
+        self
+    }
+
+    pub fn set_left_operand(&mut self, value: Operand<TS>) -> &mut ExpressionBuilder<TS> {
+        self.operands.0 = Some(value);
+        self
+    }
+
+    pub fn set_right_operand(&mut self, value: Operand<TS>) -> &mut ExpressionBuilder<TS> {
+        self.operands.1 = Some(value);
+        self
+    }
+
+    pub fn set_operator(&mut self, operator: Operator<TS>) -> &mut ExpressionBuilder<TS> {
+        self.operator = Some(operator);
+        self
+    }
+
+    fn build_function(
+        instructions: &mut Vec<Instruction<TS>>,
+        execution_context: &mut ExecutionContext<TS>,
+        function_addr: usize,
+        args: Vec<Operand<TS>>,
+        stack_size: usize,
+    ) {
+        let arg_count = args.len();
+        for arg in args {
+            match arg {
+                Operand::Function {
+                    addr,
+                    args,
+                    stack_size,
+                } => {
+                    ExpressionBuilder::build_function(
+                        instructions,
+                        execution_context,
+                        addr,
+                        args,
+                        stack_size,
+                    );
+                    instructions.push(Instruction::PushFromReturn);
+                }
+                Operand::ValueRef(addr) => instructions.push(Instruction::Push(addr)),
+                Operand::ValueRaw(val) => instructions.push(Instruction::PushRaw(val)),
+                Operand::Expression(builder) => {
+                    instructions.append(&mut builder.build(execution_context))
+                }
+            }
+        }
+        instructions.push(Instruction::Invoke(function_addr, arg_count, stack_size));
+    }
+
+    pub fn build(self, execution_context: &mut ExecutionContext<TS>) -> Vec<Instruction<TS>> {
+        let mut instructions = Vec::new();
+
+        if let Some(operand) = self.operands.0 {
```

Comment:
These are fallible, but the negative case is never checked and the function itself doesn't return anything to indicate when the operands weren't set. I would do a single if-let-else statement:

```rs
if let (Some(left), Some(right)) = self.operands {
    ...
} else {
   Err(...)
}
```

But at this point, this is barely a builder. You construct the builder and either build it immediately as a function, or set the operands and build it. Why don't we just turn these into static methods that yield `Expression` objects (which we can rename `ExpressionWriter` to) - one for making a function invocation, one for a binary operation, one for a unary operation, and one for a raw value. Then you would just compose them together using those static methods, building an expression tree and then converting it to instructions in-place.

Reply:
your right there needs to be a check but not like this, the way it is done is because you don't need to set the second operand for unary expressions
Ill add a check that the first one exists and the second one exists if the operator is a binary one

Reply:
Instead of doing it this way, I think you should just use static constructors - one for unary operations, and one for binary operations. Each would take those arguments directly, so there's no way you end up with a builder in an invalid state by accident

Reply:
well we dont need a... you mentioned this in the bottom half of your comment that I didnt read all the way through
we dont really need a builder at all this should just be 3 functions
and I was thinking we would have a macro to make building chained ones easier

Reply:
Yeah, let's convert it then. It should be simple enough to make macros for that, just wrapping some of its functions with cleaner syntax.